# **1. Overview**
Tanzania, a developing nation, faces challenges in supplying its more than 57,000,000 inhabitants with water that is safe for drinking. It is extremely difficult for people to find clean, sanitary water if they do not reside close to one of the three large lakes that border the country, as one-third of the country is arid to semi-arid. Consequently, Tanzanians rely heavily on groundwater as their primary supply of water;


According to the Sustainable Development Goals (SDG) standards, just 61% of Tanzanian households presently have access to a basic water supply, 32% to basic sanitation, and 48% to basic hygiene. As a direct result, Tanzania has had to deal with mortality and illness, with the poor and vulnerable, women, and children bearing the brunt of this burden. Inadequate WASH services are thought to be the cause of 31,000 fatalities annually in Tanzania, accounting for almost 10% of avoidable deaths. These deaths also cost the country more than $2.4 billion annually in lost productivity and additional medical expenses.

The nation already has a large number of water points (stations), but some of them require maintenance, while others have completely failed.

The project aims to builds classification model, using an iterative approach, to predict the condition of water wells in Tanzania. The dataset for modelling was obtained from data provided by Taarifa and the Tanzanian Ministry of water. 

# **2. Business and Data Understanding**
## **2.1 Business Problem**
Victoria Inc. has been procured by the Government of Tanzania on a consultancy basis to study the severe water crisis experienced by the country and propose a data driven solution to clean water accessibility. Victoria Inc. is tasked with coming up with a  model that predicts the operating condition of the water points. This model will assist the government to:
- Prioritize maintenance and repairs based on operating status;
- Understand the failure rate of the water points;
- Optimize allocation of resources to restore the water points.  

The objective of Victoria Inc is to:
- Develop a predictive model for classifying water points;
- Identify factors that affect water points functionality; 

Proposed Solution:
- Develop a machine learning classification model with a recall score of 90%.

Performance Metrics:
- Acurracy
- Precision
- Recall
- F1-score

## **2.2 Data Understanding**
###  **2.2.1 Data Source**
The dataset employed in the study was downloaded from https://www.drivendata.org/competitions/7/data/

### **2.2.2 Dataset Features**
The following features about the water points are provided:
- *amount_tsh* - Total static head (amount water available to waterpoint)
- *date_recorded* - The date the row was entered
- *funder* - Who funded the well
- *gps_height* - Altitude of the well
- *installer* - Organization that installed the well
- *longitude* - GPS coordinate
- *latitude* - GPS coordinate
- *wpt_name* - Name of the waterpoint if there is one
- *num_private* -
- *basin* - Geographic water basin
- *subvillage* - Geographic location
- *region* - Geographic location
- *region_code *- Geographic location (coded)
- *district_code* - Geographic location (coded)
- *lga* - Geographic location
- *ward* - Geographic location
- *population* - Population around the well
- *public_meeting* - True/False
- *recorded_by *- Group entering this row of data
- *scheme_management* - Who operates the waterpoint
- *scheme_name* - Who operates the waterpoint
- *permit* - If the waterpoint is permitted
- *construction_year* - Year the waterpoint was constructed
- *extraction_type* - The kind of extraction the waterpoint uses
- *extraction_type_group* - The kind of extraction the waterpoint uses
- *extraction_type_class* - The kind of extraction the waterpoint uses
- *management* - How the waterpoint is managed
- *management_group* - How the waterpoint is managed
- *payment* - What the water costs
- *payment_type* - What the water costs
- *water_quality* - The quality of the water
- *quality_group* - The quality of the water
- *quantity* - The quantity of water
- *quantity_group* - The quantity of water
- *source* - The source of the water
- *source_type* - The source of the water
- *source_class* - The source of the water
- *waterpoint_type* - The kind of waterpoint
- *waterpoint_type_group* - The kind of waterpoint

The target variable has three possible values:
- **functional** - the waterpoint is operational and there are no repairs needed
- **functional needs repair** - the waterpoint is operational, but needs repairs
- **non functional** - the waterpoint is not operational

## **2.3 Methodology**
The adopted structure for the project was CRISP-DM that entails undertaking Business Understanding; Data Understanding; Data Preparation; Data Cleaning and Explatory Data Analysis(EDA); Modelling; Conclusion and Recommendations.

# **3.Data Cleaning and EDA**
## **3.1 Data Cleaning & Preparation**
Importing packages

In [80]:
#importing standard packages
import pandas as pd 
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt 
%matplotlib inline

In [81]:
# loading the training data set
data_train_set = pd.read_csv("data/training_set_values.csv", index_col="id")
data_train_set

amount_tsh date_recorded           funder  gps_height     installer  \
id                                                                           
69572      6000.0    2011-03-14            Roman        1390         Roman   
8776          0.0    2013-03-06          Grumeti        1399       GRUMETI   
34310        25.0    2013-02-25     Lottery Club         686  World vision   
67743         0.0    2013-01-28           Unicef         263        UNICEF   
19728         0.0    2011-07-13      Action In A           0       Artisan   
...           ...           ...              ...         ...           ...   
60739        10.0    2013-05-03  Germany Republi        1210           CES   
27263      4700.0    2011-05-07      Cefa-njombe        1212          Cefa   
37057         0.0    2011-04-11              NaN           0           NaN   
31282         0.0    2011-03-08            Malec           0          Musa   
26348         0.0    2011-03-23       World Bank         191         World   

       longitude   latitude              wpt_name  num_private  \
id                                                               
69572  34.938093  -9.856322                  none            0   
8776   34.698766  -2.147466              Zahanati            0   
34310  37.460664  -3.821329           Kwa Mahundi            0   
67743  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
19728  31.130847  -1.825359               Shuleni            0   
...          ...        ...                   ...          ...   
60739  37.169807  -3.253847   Area Three Namba 27            0   
27263  35.249991  -9.070629     Kwa Yahona Kuvala            0   
37057  34.017087  -8.750434               Mashine            0   
31282  35.861315  -6.378573                Mshoro            0   
26348  38.104048  -6.747464       Kwa Mzee Lugawa            0   

                         basin  ... payment_type water_quality  quality_group  \
id                              ...                                             
69572               Lake Nyasa  ...     annually          soft           good   
8776             Lake Victoria  ...    never pay          soft           good   
34310                  Pangani  ...   per bucket          soft           good   
67743  Ruvuma / Southern Coast  ...    never pay          soft           good   
19728            Lake Victoria  ...    never pay          soft           good   
...                        ...  ...          ...           ...            ...   
60739                  Pangani  ...   per bucket          soft           good   
27263                   Rufiji  ...     annually          soft           good   
37057                   Rufiji  ...      monthly      fluoride       fluoride   
31282                   Rufiji  ...    never pay          soft           good   
26348              Wami / Ruvu  ...   on failure         salty          salty   

           quantity quantity_group                source  \
id                                                         
69572        enough         enough                spring   
8776   insufficient   insufficient  rainwater harvesting   
34310        enough         enough                   dam   
67743           dry            dry           machine dbh   
19728      seasonal       seasonal  rainwater harvesting   
...             ...            ...                   ...   
60739        enough         enough                spring   
27263        enough         enough                 river   
37057        enough         enough           machine dbh   
31282  insufficient   insufficient          shallow well   
26348        enough         enough          shallow well   

                source_type source_class              waterpoint_type  \
id                                                                      
69572                spring  groundwater           communal standpipe   
8776   rainwater harvesting      surface           communal standpipe   
34310 

In [82]:
# importing the test data set
data_test_set = pd.read_csv("data/test_set_values.csv", index_col="id")
data_test_set

amount_tsh date_recorded                  funder  gps_height  \
id                                                                    
50785         0.0    2013-02-04                    Dmdd        1996   
51630         0.0    2013-02-04  Government Of Tanzania        1569   
17168         0.0    2013-02-01                     NaN        1567   
45559         0.0    2013-01-22              Finn Water         267   
49871       500.0    2013-03-27                  Bruder        1260   
...           ...           ...                     ...         ...   
39307         0.0    2011-02-24                  Danida          34   
18990      1000.0    2011-03-21                    Hiap           0   
28749         0.0    2013-03-04                     NaN        1476   
33492         0.0    2013-02-18                 Germany         998   
68707         0.0    2013-02-13  Government Of Tanzania         481   

        installer  longitude   latitude                 wpt_name  num_private  \
id                                                                              
50785        DMDD  35.290799  -4.059696  Dinamu Secondary School            0   
51630         DWE  36.656709  -3.309214                  Kimnyak            0   
17168         NaN  34.767863  -5.004344           Puma Secondary            0   
45559  FINN WATER  38.058046  -9.418672           Kwa Mzee Pange            0   
49871      BRUDER  35.006123 -10.950412          Kwa Mzee Turuka            0   
...           ...        ...        ...                      ...          ...   
39307          Da  38.852669  -6.582841                Kwambwezi            0   
18990        HIAP  37.451633  -5.350428         Bonde La Mkondoa            0   
28749         NaN  34.739804  -4.585587                  Bwawani            0   
33492         DWE  35.432732 -10.584159                 Kwa John            0   
68707  Government  34.765054 -11.226012         Kwa Mzee Chagala            0   

                         basin  ... payment_type water_quality  quality_group  \
id                              ...                                             
50785                 Internal  ...    never pay          soft           good   
51630                  Pangani  ...    never pay          soft           good   
17168                 Internal  ...    never pay          soft           good   
45559  Ruvuma / Southern Coast  ...      unknown          soft           good   
49871  Ruvuma / Southern Coast  ...      monthly          soft           good   
...                        ...  ...          ...           ...            ...   
39307              Wami / Ruvu  ...    never pay          soft           good   
18990                  Pangani  ...     annually         salty          salty   
28749                 Internal  ...    never pay          soft           good   
33492               Lake Nyasa  ...    never pay          soft           good   
68707               Lake Nyasa  ...    never pay          soft           good   

           quantity quantity_group                source  \
id                                                         
50785      seasonal       seasonal  rainwater harvesting   
51630  insufficient   insufficient                spring   
17168  insufficient   insufficient  rainwater harvesting   
45559           dry            dry          shallow well   
49871        enough         enough                spring   
...             ...            ...                   ...   
39307        enough         enough                 river   
18990  insufficient   insufficient          shallow well   
28749  insufficient   insufficient                   dam   
33492  insufficient   insufficient                 river   
68707           dry            dry                spring   

                source_type source_class     waterpoint_type  \
id                                                             
50785  rainwater harvesting      surface               other   
51

In [83]:
# importing the training set labels
data_train_labels = pd.read_csv("data/training_set_labels.csv", index_col="id")
data_train_labels

status_group
id                   
69572      functional
8776       functional
34310      functional
67743  non functional
19728      functional
...               ...
60739      functional
27263      functional
37057      functional
31282      functional
26348      functional

[59400 rows x 1 columns]

In [84]:
#merging the training dataset with the train labels
data = pd.merge(data_train_labels, data_train_set, on="id", how="inner")
data

status_group  amount_tsh date_recorded           funder  gps_height  \
id                                                                             
69572      functional      6000.0    2011-03-14            Roman        1390   
8776       functional         0.0    2013-03-06          Grumeti        1399   
34310      functional        25.0    2013-02-25     Lottery Club         686   
67743  non functional         0.0    2013-01-28           Unicef         263   
19728      functional         0.0    2011-07-13      Action In A           0   
...               ...         ...           ...              ...         ...   
60739      functional        10.0    2013-05-03  Germany Republi        1210   
27263      functional      4700.0    2011-05-07      Cefa-njombe        1212   
37057      functional         0.0    2011-04-11              NaN           0   
31282      functional         0.0    2011-03-08            Malec           0   
26348      functional         0.0    2011-03-23       World Bank         191   

          installer  longitude   latitude              wpt_name  num_private  \
id                                                                             
69572         Roman  34.938093  -9.856322                  none            0   
8776        GRUMETI  34.698766  -2.147466              Zahanati            0   
34310  World vision  37.460664  -3.821329           Kwa Mahundi            0   
67743        UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
19728       Artisan  31.130847  -1.825359               Shuleni            0   
...             ...        ...        ...                   ...          ...   
60739           CES  37.169807  -3.253847   Area Three Namba 27            0   
27263          Cefa  35.249991  -9.070629     Kwa Yahona Kuvala            0   
37057           NaN  34.017087  -8.750434               Mashine            0   
31282          Musa  35.861315  -6.378573                Mshoro            0   
26348         World  38.104048  -6.747464       Kwa Mzee Lugawa            0   

       ... payment_type water_quality quality_group      quantity  \
id     ...                                                          
69572  ...     annually          soft          good        enough   
8776   ...    never pay          soft          good  insufficient   
34310  ...   per bucket          soft          good        enough   
67743  ...    never pay          soft          good           dry   
19728  ...    never pay          soft          good      seasonal   
...    ...          ...           ...           ...           ...   
60739  ...   per bucket          soft          good        enough   
27263  ...     annually          soft          good        enough   
37057  ...      monthly      fluoride      fluoride        enough   
31282  ...    never pay          soft          good  insufficient   
26348  ...   on failure         salty         salty        enough   

       quantity_group                source           source_type  \
id                                                                  
69572          enough                spring                spring   
8776     insufficient  rainwater harvesting  rainwater harvesting   
34310          enough                   dam                   dam   
67743             dry           machine dbh              borehole   
19728        seasonal  rainwater harvesting  rainwater harvesting   
...               ...                   ...                   ...   
60739          enough                spring                spring   
27263          enough                 river            river/lake   
37057          enough           machine dbh              borehole   
31282    insufficient          shallow well          shallow well   
26348          enough          shallow well          shallow well   

       source_class              waterpoint_type waterpoint_type_group  
id                                                        

In [85]:
# creating new index for the merged dataset
data.reset_index(inplace=True)
data

id    status_group  amount_tsh date_recorded           funder  \
0      69572      functional      6000.0    2011-03-14            Roman   
1       8776      functional         0.0    2013-03-06          Grumeti   
2      34310      functional        25.0    2013-02-25     Lottery Club   
3      67743  non functional         0.0    2013-01-28           Unicef   
4      19728      functional         0.0    2011-07-13      Action In A   
...      ...             ...         ...           ...              ...   
59395  60739      functional        10.0    2013-05-03  Germany Republi   
59396  27263      functional      4700.0    2011-05-07      Cefa-njombe   
59397  37057      functional         0.0    2011-04-11              NaN   
59398  31282      functional         0.0    2011-03-08            Malec   
59399  26348      functional         0.0    2011-03-23       World Bank   

       gps_height     installer  longitude   latitude              wpt_name  \
0            1390         Roman  34.938093  -9.856322                  none   
1            1399       GRUMETI  34.698766  -2.147466              Zahanati   
2             686  World vision  37.460664  -3.821329           Kwa Mahundi   
3             263        UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu   
4               0       Artisan  31.130847  -1.825359               Shuleni   
...           ...           ...        ...        ...                   ...   
59395        1210           CES  37.169807  -3.253847   Area Three Namba 27   
59396        1212          Cefa  35.249991  -9.070629     Kwa Yahona Kuvala   
59397           0           NaN  34.017087  -8.750434               Mashine   
59398           0          Musa  35.861315  -6.378573                Mshoro   
59399         191         World  38.104048  -6.747464       Kwa Mzee Lugawa   

       ...  payment_type water_quality quality_group      quantity  \
0      ...      annually          soft          good        enough   
1      ...     never pay          soft          good  insufficient   
2      ...    per bucket          soft          good        enough   
3      ...     never pay          soft          good           dry   
4      ...     never pay          soft          good      seasonal   
...    ...           ...           ...           ...           ...   
59395  ...    per bucket          soft          good        enough   
59396  ...      annually          soft          good        enough   
59397  ...       monthly      fluoride      fluoride        enough   
59398  ...     never pay          soft          good  insufficient   
59399  ...    on failure         salty         salty        enough   

       quantity_group                source           source_type  \
0              enough                spring                spring   
1        insufficient  rainwater harvesting  rainwater harvesting   
2              enough                   dam                   dam   
3                 dry           machine dbh              borehole   
4            seasonal  rainwater harvesting  rainwater harvesting   
...               ...                   ...                   ...   
59395          enough                spring                spring   
59396          enough                 river            river/lake   
59397          enough           machine dbh              borehole   
59398    insufficient          shallow well          shallow well   
59399          enough          shallow well          shallow well   

      source_class              waterpoint_type waterpoint_type_group  
0      groundwater           communal standpipe    communal standpipe  
1          surface           communal standpipe    communal standpipe  
2          surface  communal standpipe multiple    communal standpipe  
3      groundwater  communal standpipe multiple    communal standpipe  
4          surface           communal standpipe    communal standpipe  
...            ...                          ...           

In [86]:
# checking the datatypes
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   status_group           59400 non-null  object 
 2   amount_tsh             59400 non-null  float64
 3   date_recorded          59400 non-null  object 
 4   funder                 55765 non-null  object 
 5   gps_height             59400 non-null  int64  
 6   installer              55745 non-null  object 
 7   longitude              59400 non-null  float64
 8   latitude               59400 non-null  float64
 9   wpt_name               59400 non-null  object 
 10  num_private            59400 non-null  int64  
 11  basin                  59400 non-null  object 
 12  subvillage             59029 non-null  object 
 13  region                 59400 non-null  object 
 14  region_code            59400 non-null  int64  
 15  di

The data set is made up of 41 columns and 59,400 rows. The data frame features has 3 float datatype; 7 integer datatype and 31 object datatype.

In [87]:
#checking for duplicates
data.duplicated().sum()

0

The data set had zero duplicates

In [88]:
# checking null values
data.isna().sum()

id                           0
status_group                 0
amount_tsh                   0
date_recorded                0
funder                    3635
gps_height                   0
installer                 3655
longitude                    0
latitude                     0
wpt_name                     0
num_private                  0
basin                        0
subvillage                 371
region                       0
region_code                  0
district_code                0
lga                          0
ward                         0
population                   0
public_meeting            3334
recorded_by                  0
scheme_management         3877
scheme_name              28166
permit                    3056
construction_year            0
extraction_type              0
extraction_type_group        0
extraction_type_class        0
management                   0
management_group             0
payment                      0
payment_type                 0
water_qu

The columns "funder", "installer", "subvillage", "public_meeting", "scheme_management", "scheme_name", and "permit" had 3635, 3655, 371, 3334, 3877, 28166 and 3056 missing (null) values respectively. Scheme_name has the highest number of missing values.

In [89]:
for col in data.columns:
    print(data[col].value_counts())

69572    1
27851    1
6924     1
61097    1
48517    1
        ..
59036    1
56446    1
3855     1
52786    1
26348    1
Name: id, Length: 59400, dtype: int64
functional                 32259
non functional             22824
functional needs repair     4317
Name: status_group, dtype: int64
0.0         41639
500.0        3102
50.0         2472
1000.0       1488
20.0         1463
            ...  
6300.0          1
120000.0        1
138000.0        1
350000.0        1
59.0            1
Name: amount_tsh, Length: 98, dtype: int64
2011-03-15    572
2011-03-17    558
2013-02-03    546
2011-03-14    520
2011-03-16    513
             ... 
2011-09-11      1
2011-08-31      1
2011-09-21      1
2011-08-30      1
2013-12-01      1
Name: date_recorded, Length: 356, dtype: int64
Government Of Tanzania    9084
Danida                    3114
Hesawa                    2202
Rwssp                     1374
World Bank                1349
                          ... 
Rarymond Ekura               1
Justin

The following columns contain either similar or duplicated data, therefore, in order to avoid multicollinearity one or both of the columns will be dropped:
- *scheme_management *and *management*;
- *extraction_type*, *extraction_type_group* and *extraction_type_class*;
- *payment* and *payment_type*;
- *water_quality* and *quality_group*
- *quantity* and *quantity_group*
- *source* and *source_type*
- *waterpoint_type* and *waterpoint_type_group*

Similarly, the following columns will either be transformed or dropped: 
- columns to be dropped: 
    - *id* as its just an index identifier, *num_private* since its not defined and therefore its relevance is not clear; *recorded_by* since it has the same value throughout the dataset; 
    - *population*, *amount_tsh*, *construction_year*, *longitude*, *latitude*, and *gps_height* have 0 entered on most of their rows. The rows with 0 will be dropped.
- columns to be transformed or feature engineered:
    - *permit* and *public_meeting* are boolean;
    - *wpt_name*, *scheme_name*, *subvillage*, *ward*, *date_recorded*, *funder* are categorical variables with their unique values in integers. Therefore, they will require further analysis and probable use of dummy variables.

In [90]:
# columns to be dropped
drop_cols = ["scheme_management", "extraction_type", "extraction_type_group", 
             "payment", "water_quality", "source_type", "waterpoint_type_group", 
             "id", "num_private", "recorded_by", "quantity", "public_meeting", 
             "wpt_name", "scheme_name", "subvillage", "ward", "date_recorded", 
             "funder", "district_code", "lga", "region_code", "ward", "management_group", 
             "longitude", "latitude", "gps_height", "source"]

#dropping the columns:
data = data.drop(drop_cols, axis=1)

# checking the information of the new dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   status_group           59400 non-null  object 
 1   amount_tsh             59400 non-null  float64
 2   installer              55745 non-null  object 
 3   basin                  59400 non-null  object 
 4   region                 59400 non-null  object 
 5   population             59400 non-null  int64  
 6   permit                 56344 non-null  object 
 7   construction_year      59400 non-null  int64  
 8   extraction_type_class  59400 non-null  object 
 9   management             59400 non-null  object 
 10  payment_type           59400 non-null  object 
 11  quality_group          59400 non-null  object 
 12  quantity_group         59400 non-null  object 
 13  source_class           59400 non-null  object 
 14  waterpoint_type        59400 non-null  object 
dtypes:

The new dataframe has 15 columns and 59400 rows. Its only columns "installer" and "permit" that have missing values. The dataframe is made up of three datatypes: 1 column of type float, 2 columns with type integer and 12 columns with type object.

In [91]:
# dropping the null values
data = data.dropna()

# confirming no null values are remaining
data.isna().sum()

status_group             0
amount_tsh               0
installer                0
basin                    0
region                   0
population               0
permit                   0
construction_year        0
extraction_type_class    0
management               0
payment_type             0
quality_group            0
quantity_group           0
source_class             0
waterpoint_type          0
dtype: int64

In [92]:
# converting permit column to integers
data["permit"] = data["permit"].astype(int)

# checking datatypes of the dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55102 entries, 0 to 59399
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   status_group           55102 non-null  object 
 1   amount_tsh             55102 non-null  float64
 2   installer              55102 non-null  object 
 3   basin                  55102 non-null  object 
 4   region                 55102 non-null  object 
 5   population             55102 non-null  int64  
 6   permit                 55102 non-null  int32  
 7   construction_year      55102 non-null  int64  
 8   extraction_type_class  55102 non-null  object 
 9   management             55102 non-null  object 
 10  payment_type           55102 non-null  object 
 11  quality_group          55102 non-null  object 
 12  quantity_group         55102 non-null  object 
 13  source_class           55102 non-null  object 
 14  waterpoint_type        55102 non-null  object 
dtypes:

The dataframe shape has been reduced to 15 columns by 55102 rows.

In [93]:
# checking unique values after initial data cleaning
data.nunique()

status_group                3
amount_tsh                 95
installer                2056
basin                       9
region                     21
population               1026
permit                      2
construction_year          55
extraction_type_class       7
management                 12
payment_type                7
quality_group               6
quantity_group              5
source_class                3
waterpoint_type             7
dtype: int64

Categorical variable "installer", requires further investigation since it still has 2056 unique variables which would overwhelm the model. Population is an integer and therefore having different unique values is not problematic.

In [94]:
# viewing the installer column data values
print(data["installer"].unique().tolist())


['Roman', 'GRUMETI', 'World vision', 'UNICEF', 'Artisan', 'DWE', 'DWSP', 'Water Aid', 'Private', 'DANIDA', 'Lawatefuka water sup', 'WEDECO', 'Danid', 'TWE', 'ISF', 'Kilolo Star', 'District council', 'Water', 'WU', 'Not known', 'Central government', 'CEFA', 'Commu', 'Accra', 'World Vision', 'LGA', 'MUWSA', 'KKKT _ Konde and DWE', 'Government', 'Olgilai village community', 'KKKT', 'RWE', 'Adra /Community', 'SEMA', 'SHIPO', 'HESAWA', 'ACRA', 'Community', 'IFAD', 'Sengerema Water Department', 'HE', 'ISF and TACARE', 'Kokeni', 'DA', 'Adra', 'ALLYS', 'AICT', 'KIUMA', 'CES', 'District Counci', 'Ruthe', 'Adra/Community', 'Tulawaka Gold Mine', 'KKT C', 'Water board', 'LOCAL CONTRACT', 'LIPS', 'TASAF', 'World', '0', 'SW', 'Shipo', 'Fini water', 'Kanisa', 'OXFARM', 'VILLAGE COUNCIL Orpha', 'Villagers', 'Idara ya maji', 'FPCT', 'WVT', 'Ir', 'DANID', 'Angli', 'secondary school', 'Amref', 'JBG', 'DADIS', 'International Aid Services', 'RW', 'Dmdd', 'TCRS', 'RC Church', 'WATER AID', 'JICA', 'Gwasco L'

In [95]:
#exporting installer column details to a file
data['installer'].to_csv('data/column_details.csv', index=False)


In [96]:
# correcting typos on the installer column

data["installer"] = data["installer"].replace(to_replace=(
    'World vision', 'World Vision', 'world vision', 'World Visiin', 'World Division', 
    'WORLD VISION'), value="World Vision")

data["installer"] = data["installer"].replace(to_replace=(
    'Central government', 'Government', 'Central Government','GOVERNMENT', 'Tanzania Government', 
    'CENTRAL GOVERNMENT', 'Government /Community', 'Concern /government', 'central government',
    'Cental Government','Cebtral Government', 'TANZANIAN GOVERNMENT', 'Tanzanian Government',
    'Government and Community', 'Cetral government /RC', 'Tanzania government', 'Centra Government',
    'GOVERNME', 'GOVER', 'Gover','Central govt', 'Gove', 'GOVERM', 'GOVERN',
    'Centra govt', 'Cetral government', 'Governmen', 'TZ as', 'Jumuiya', 'Jumuhia', ), 
                                              value="Central Government")

data["installer"] = data["installer"].replace(to_replace=(
    'World Bank', 'World banks', 'WORLD BANK', 'World bank','WOULD BANK', 'Word Bank',
    'world banks', 'WORDL BANK', 'Word bank', ), 
                                              value="World Bank")

data["installer"] = data["installer"].replace(to_replace=('UNICRF', 'Unicef', 'Unisef'), 
                                              value="UNICEF")

data["installer"] = data["installer"].replace(to_replace=(
    'JICA', 'JIKA', 'Jika', 'Jica', 'Jiks', 'JAICA','Jaica', 'JAICA CO', 'Jeica', 'GAICA'), 
                                              value="JICA")

data["installer"] = data["installer"].replace(to_replace=(
    'Olgilai village community', 'VILLAGE COUNCIL Orpha', 'Villagers', 'Villa', 
    'Villages', 'Vill', 'Village', 'VILLAGE COUNCIL', 'Villi', 'Village Council', 
    'Village Counil', 'Sekei village community', 'Village govt', 'Village community', 
    'Villaers', 'Village Govt', 'ViLLAGE COUNCIL', 'villagers', 'Village Government', 
     'villigers', 'VILLAGE COUNCIL .ODA','VILLAG', 'Villag', 'Villege Council', 
     'Villagerd','VILLAGERS', 'Nduku village','Subvillage', 'Kitiangare village community', 
     'Village Technician', 'Oldadai village community', 'VILLAGE WATER COMMISSION', 
     'villager','Village Office', 'Village community members', 'Nyamingu subvillage',
     'Village water committee', 'Village water attendant', 'Village local contractor'), 
                                              value="VILLAGE")

data["installer"] = data["installer"].replace(to_replace="0", value="Unknown")

data["installer"] = data["installer"].replace(to_replace=(
    'Adra/ Community',  'Arab community', 'Taboma/Community','Communit', 'Killflora /Community',
    'RWE/ Community', 'Olgilai village community', 'Commu', 'Sekei village community', 'COMMU',
    'COMMUNITY', 'Ilwilo community', 'Igolola community', 'Comunity', 'Marumbo Community', 
    'Maseka community','Islamic community', 'District Community j' ), 
                                              value="COMMUNITY")

data["installer"] = data["installer"].replace(to_replace=(
    'Fini water', 'FinW', 'FIN WATER','FINN WATER', 'FINI WATER', 'FinWater', 
    'FiNI WATER', 'FINI Water', 'Fin water', 'FinWate', ), 
                                              value="FINI WATER")

data["installer"] = data["installer"].replace(to_replace=(
    'District council','District Counci' 'District Council','District water department', 
    'Distri', 'District Water Department', 'District water depar', 'District COUNCIL',
    'DISTRICT COUNCIL', 'MBULU DISTRICT COUNCIL', 'Sangea District Coun', 
    'Songea District Coun', 'Mbozi District Council'), 
                                              value="DISTRICT COUNCIL")

data["installer"] = data["installer"].replace(to_replace=(
    'RC Church', 'RC CHURCH','RC Churc', 'RCchurch/CEFA', 'Chur',
    'RC church/Central Gover', 'KKKT CHURCH', 'Pentecost church', 'Roman Church', 
    'RC church/CEFA', 'lutheran church', 'Free Pentecoste Church of Tanz', 'RC C', 
    'Church', 'Morovian church', 'CEFA/rc church', 'Anglican Church', 'Church Of Disciples', 
    'ANGLIKANA CHURCH', 'ANGRIKANA', 'church', 'Anglican church', 'kanisa', 
    'Lutheran Church', 'Pentecostal church','Jeshi la wokovu [cida]', 'METHODIST CHURCH',
    'CHURCH', 'morovian church', 'Angrikana', 'HAIDOMU LUTHERAN CHURCH','RC CHURCH BROTHER', 
    'St Magreth Church', 'Pentecosta','rc church', 'Anglica Church', 'RC Mi', 
    'BAPTIST CHURCH OF TANZANIA', 'FPCT Church', 'RC .Church','BATIST CHURCH', 'CHURC', 
    'DWE/Anglican church', 'RC Churc','RCchurch/CEFA', 'RC', 'RC Ch','HW/RC' 'RC CH',
    'rc ch','RC CATHORIC', 'RC mission',  'Church Of Disciples', 'CG/RC', 'RC MISSIONARY', 'RC Msufi', 
    'Rc Mis','Jeshi la Wokovu'), 
                                              value="CHURCH")

data["installer"] = data["installer"].replace(to_replace=(
    'DANIDA', 'Danid', 'DANID', 'DANIDA CO','DANIAD', 'DANIDS' ), 
                                              value="DANIDA")

data["installer"] = data["installer"].replace(to_replace=(
    'HESAWA', 'Hesawa', 'HESAWS',  'hesawa', 'HESAW', 'HESAWZ', 'HESAWQ', 'Hesawz', 'HesaWa',   ), 
                                              value="HESAWA")

data["installer"] = data["installer"].replace(to_replace=(
    'KKKT _ Konde and DWE','KKKT','KKKT-Dioces ya Pare', 'KkKT', 'KKKT Leguruki', 
    'KKKT Ndrumangeni','KKKT DME', 'KKKT Kilinga', 'KKKT Canal', 'KKKT Katiti juu', 
    'KKKT MAREU', 'Kkkt'), 
                                              value="KKKT")
# viewing the top 30 installer values after initial cleaning
data["installer"].value_counts().head(30)

DWE                   17361
Central Government     3587
DANIDA                 1676
HESAWA                 1225
RWE                    1203
COMMUNITY              1162
KKKT                   1091
DISTRICT COUNCIL        817
Unknown                 780
TCRS                    702
World Vision            671
CHURCH                  647
CES                     610
FINI WATER              572
Community               552
District Council        546
VILLAGE                 514
JICA                    427
LGA                     408
WEDECO                  397
TASAF                   390
UNICEF                  333
TWESA                   316
AMREF                   313
WU                      301
Dmdd                    286
ACRA                    277
SEMA                    249
DW                      246
OXFAM                   234
Name: installer, dtype: int64

## **3.2 EDA**

# **4. Modelling**

# **5. Evaluation**

# **6. Conclusion and Recommendation**
## **6.1 Conclusion**

## **6.2 Recommendations**